crates/rolldown/src/module_finalizers/mod.rs

```diff
@@ -310,7 +310,28 @@ impl<'me, 'ast> ScopeHoistingFinalizer<'me, 'ast> {
       )),
     );
 
-    let exports_len = self.ctx.linking_info.canonical_exports(false).count();
+    // Collect exports whose declaring statements survive tree-shaking. This ensures that we don't
+    // generate getters for bindings that were already removed, which would otherwise keep them
+    // alive when HMR is enabled.
+    let included_exports: Vec<_> = self
+      .ctx
+      .linking_info
+      .canonical_exports(false)
+      .filter(|(_name, resolved_export)| {
+        let owner_idx = resolved_export.symbol_ref.owner;
+        match &self.ctx.modules[owner_idx] {
+          Module::Normal(m) => m
+            .stmt_infos
+            .declared_stmts_by_symbol(&resolved_export.symbol_ref)
+            .iter()
+            .any(|id| m.stmt_infos[*id].is_included),
```

Comment:
I'm worried if this won't affect performance. Maybe the filtering needs to be done if only HMR is enabled?